src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs

```diff
@@ -0,0 +1,387 @@
+            if (settings.EnableSkipToken)
+            {
+                if (expandedItem != null)
+                {
+                    if (expandedItem.OrderByOption != null)
+                    {
+                        orderByNodes = OrderByNode.CreateCollection(expandedItem.OrderByOption);
+                    }
+
+                    skipTokenGenerator = (obj) =>
+                    {
+                        return GenerateSkipTokenValue(obj, model, orderByNodes);
+                    };
+                    
+                    return GetNextPageHelper.GetNextPageLink(baseUri, pageSize, instance, skipTokenGenerator);
+                }
+
+                if (context.QueryOptions != null && context.QueryOptions.OrderBy != null)
+                {
+                    orderByNodes = context.QueryOptions.OrderBy.OrderByNodes;
+                }
+
+                skipTokenGenerator = (obj) =>
+                {
+                    return GenerateSkipTokenValue(obj, model, orderByNodes);
+                };
+            }
+            return context.InternalRequest.GetNextPageLink(pageSize, instance, skipTokenGenerator);
+        }
+
+        /// <summary>
+        /// Returns a function that converts an object to a skiptoken value string
+        /// </summary>
+        /// <param name="lastMember"> Object based on which SkipToken value will be generated.</param>
+        /// <param name="model">The edm model.</param>
+        /// <param name="orderByNodes">QueryOption </param>
+        /// <returns></returns>
+        protected virtual string GenerateSkipTokenValue(Object lastMember, IEdmModel model, IList<OrderByNode> orderByNodes)
+        {
+            object value;
+            if (lastMember == null)
+            {
+                return String.Empty;
+            }
+            IEnumerable<IEdmProperty> propertiesForSkipToken = GetPropertiesForSkipToken(lastMember, model, orderByNodes);
+
+            String skipTokenvalue = String.Empty;
+            if (propertiesForSkipToken == null)
+            {
+                return skipTokenvalue;
+            }
+
+            int count = 0;
+            int lastIndex = propertiesForSkipToken.Count() - 1;
+            foreach (IEdmProperty property in propertiesForSkipToken)
+            {
+                bool islast = count == lastIndex;
+                string propertyName = property.Name;
+                IEdmStructuredObject obj = lastMember as IEdmStructuredObject;
+                if (obj != null)
+                {
+                    obj.TryGetPropertyValue(property.Name, out value);
+                }
+                else
+                {
+                    value = lastMember.GetType().GetProperty(property.Name).GetValue(lastMember);
+                }
+
+                String uriLiteral = String.Empty;
+                if (value == null)
+                {
+                    uriLiteral = ODataUriUtils.ConvertToUriLiteral(value, ODataVersion.V401);
```

Comment:
>uriLiteral = ODataUriUtils.ConvertToUriLiteral(value, ODataVersion.V401); [](start = 20, length = 73)

If value is null, can we just use "null" and not call ConvertToUriLiteral?

Reply:
It is required because cannot be serialized back from null to string using ConvertFromUriLiteral. 

---
In reply to: [282594775](https://github.com/OData/WebApi/pull/1724#discussion_r282594775) [](ancestors = 282594775)